### DISCUSSION TITLE
Typing support

### DISCUSSION THREAD

**Discussion by nameloCmaS** [hyperdiv/hyperdiv#69] [Ideas]
Hello!

I am planning on adding typing support to make it easier to code in the framework.

The example apps have lots of static typing issues which could deter people from using this framework over another.

I have added typing to the `Prop` class and its descriptors, which removed some of these errors. I have since added to a few more classes.

Will you be happy if I submit these in small PR batches?

**Comment by mariusnita** [comment]
I'm interested in understanding this better. Can you expand on what static typing issues you're seeing, which IDE, type checker, etc.?

And also I'm interested to see what these typing annotations look like, especially for props, some of which accept fairly complex unions of value shapes. E.g. `Prop(Size)` accepts a string, a 2-tuple (whose second component is a union of constants -- "px", "em", "rem", etc.), a number, etc.

Thanks!

**Comment by idvorkin** [comment]
I've been thinking about this a bit too. A few pointers if it's helpful: 

1/ Using Pydantic class to define state (and have it auto translate to props)
2/ Using [Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated) , Typer has nice ergonomics for [this](https://typer.tiangolo.com/tutorial/arguments/optional/)

**Comment by nameloCmaS** [comment]
I got about as far as adding annotations to most of the frame classes, props, collector, component, and hyperdivtype code.

This removed most of the static type hint warnings for the 'user' demo app using PyRight; the parse function of the props was managed but may be a bit of a hack (I think I used a TypeVar, but PyRight didn't like this as it is the incorrect use; it doesn't quite work either). It also removed some of the hinting that was inferred such as the forms; it couldn't work out the function parameters any longer.

To make it work properly I think the prop class would need a metaclass to allow the different typing for the class and instance variables. This would be fairly simple, but not tested.

Similar possibly for the collector but as more backend, may not be needed. As components inherit it though, could be needed.

The way the props are used for the different component initialisations and their mixin classes will need some more thought on how these could be type hinted.
